gym/vector/async_vector_env.py

```diff
@@ -0,0 +1,482 @@
+
+        if self.waiting_step:
+            logger.error('Calling `step_async` while waiting for a pending '
+                'call to `step` to complete. Closing `{0}`...'.format(
+                type(self).__name__))
+            self.close(terminate=True)
+            raise AlreadySteppingError('Calling `step_async` while waiting for '
+                'a pending call to `step` to complete.')
+
+        for pipe, action in zip(self.parent_pipes, actions):
+            pipe.send(('step', action))
+        self.waiting_step = True
+
+    def step_wait(self, timeout=None):
+        """
+        Parameters
+        ----------
+        timeout : int, optional
+            Number of seconds before the call to `step_wait` times out. If
+            `None`, the call to `step_wait` never times out.
+
+        Returns
+        -------
+        observations : sample from `observation_space`
+            A batch of observations from the vectorized environment.
+
+        rewards : `np.ndarray` instance (dtype `np.float_`)
+            A vector of rewards from the vectorized environment.
+
+        dones : `np.ndarray` instance (dtype `np.bool_`)
+            A vector whose entries indicate whether the episode has ended.
+
+        infos : list of dict
+            A list of auxiliary diagnostic informations.
+        """
+        self._assert_is_running()
+        if self.waiting_reset or not self.waiting_step:
+            logger.error('Calling `step_wait` without any prior call to '
+                '`step_async`. Closing `{0}`...'.format(type(self).__name__))
+            self.close(terminate=True)
+            raise NotSteppingError('Calling `step_wait` without any prior call '
+                'to `step_async`.')
+
+        if not self.poll(timeout):
+            self.waiting_step = False
+            raise mp.TimeoutError('The call to `step_wait` has timed out after '
+                '{0} second{1}.'.format(timeout, 's' if timeout > 1 else ''))
+
+        self._restart_if_errors(reset=False)
+        results = [pipe.recv() for pipe in self.parent_pipes]
+        self.waiting_step = False
+        observations_list, rewards, dones, infos = zip(*results)
+
+        if not self.shared_memory:
+            concatenate(observations_list, self.observations,
+                self.single_observation_space)
+
+        return (deepcopy(self.observations) if self.copy else self.observations,
+                np.array(rewards), np.array(dones, dtype=np.bool_), infos)
+
+    def poll(self, timeout=None):
+        self._assert_is_running()
+        if timeout is not None:
+            end_time = time.time() + timeout
+        delta = None
+        for pipe in self.parent_pipes:
+            if timeout is not None:
+                delta = max(end_time - time.time(), 0)
+            if not pipe.poll(delta):
+                break
+        else:
+            return True
+        return False
+
+    def close_extras(self, timeout=None, terminate=False):
```

Comment:
The existing subprocvecenv doesn't seem to need the `terminate` or `timeout` arguments.  While I could see an argument for the timeout one, why add `terminate`?

Reply:
Actually, what's the argument for timeout as well?  We haven't needed it in baselines.

Reply:
The idea behind `timeout` is to leave `close` the possibility to close the individual environment gracefully, but doesn't have to wait for too long (for example, if something went wrong in one of the individual environment that leaves it hanging forever). `terminate` is the extreme version of it, where the user wants to close the environment without having to wait at all.

Reply:
Sorry, that wasn't a very clear question.  I realize why they're there, it's just I haven't used an environment where they were necessary.  It seems like something that would be handled by a wrapper instead of the environment vectorizer.  Do you have a case where this is actually used?

Reply:
I never had to use it, but I feel like leaving the user the ability to have a finer control over this if they need to might be a good idea. I think it makes sense to have it in the base class for `AsyncVectorEnv`, especially for handling termination of processes, but I'm completely open to a wrapper-based solution if you have an idea.